### PR TITLE
Make the scheme optional for domain and topLevelDomain function

### PR DESCRIPTION
**Pull request by YiuRULE**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Category (leave one):
- Bug Fix

Short description (up to few sentences):

The functions `domain` and `topLevelDomain` returned an empty string when we send an URL without a scheme. For example send 'www.yandex.com' or 'yandex.com' were not valid argument and returned an empty string. So user didn't have any choice to do an external operation with adding `//` or a full protocol.

### CODE REVIEW

File: dbms/src/Functions/domain.h [hunk 1]
```diff
@@ -9,13 +9,19 @@ namespace DB
...
 inline StringRef getURLHost(const char * data, size_t size)
 {
     Pos pos = data;
     Pos end = data + size;
 
     if (end == (pos = find_first_symbols<'/'>(pos, end)))
-        return {};
+    {
+        if (ignore_scheme)
+            pos = data;
```
Comment (alexey-milovidov) [inline dbms/src/Functions/domain.h]: The code became less readable. Better to introduce additional variable `slash_pos`...

File: dbms/src/Functions/domain.h [hunk 2]
```diff
@@ -9,13 +9,19 @@ namespace DB
 {
 
 /// Extracts host from given url.
+template <bool ignore_scheme = true>
```
Comment (alexey-milovidov) [inline dbms/src/Functions/domain.h]: As we use only single variant of this function, template is unnecessary.

**Comment by alexey-milovidov** [comment]
Let's wait for performance tests to run (we have `performance/url/url_hits.xml` for these functions).

**Comment by alexey-milovidov** [comment]
Stateful tests should be updated. You can download test data according to instruction:

```
sudo apt install wget xz-utils

wget https://clickhouse-datasets.s3.yandex.net/hits/tsv/hits_v1.tsv.xz
wget https://clickhouse-datasets.s3.yandex.net/visits/tsv/visits_v1.tsv.xz

xz -v -d hits_v1.tsv.xz
xz -v -d visits_v1.tsv.xz

clickhouse-client

CREATE TABLE test.hits ( WatchID UInt64,  JavaEnable UInt8,  Title String,  GoodEvent Int16,  EventTime DateTime,  EventDate Date,  CounterID UInt32,  ClientIP UInt32,  ClientIP6 FixedString(16),  RegionID UInt32,  UserID UInt64,  CounterClass Int8,  OS UInt8,  UserAgent UInt8,  URL String,  Referer String,  URLDomain String,  RefererDomain String,  Refresh UInt8,  IsRobot UInt8,  RefererCategories Array(UInt16),  URLCategories Array(UInt16),  URLRegions Array(UInt32),  RefererRegions Array(UInt32),  ResolutionWidth UInt16,  ResolutionHeight UInt16,  ResolutionDepth UInt8,  FlashMajor UInt8,  FlashMinor UInt8,  FlashMinor2 String,  NetMajor UInt8,  NetMinor UInt8,  UserAgentMajor UInt16,  UserAgentMinor FixedString(2),  CookieEnable UInt8,  JavascriptEnable UInt8,  IsMobile UInt8,  MobilePhone UInt8,  MobilePhoneModel String,  Params String,  IPNetworkID UInt32,  TraficSourceID Int8,  SearchEngineID UInt16,  SearchPhrase String,  AdvEngineID UInt8,  IsArtifical UInt8,  WindowClientWidth UInt16,  WindowClientHeight UInt16,  ClientTimeZone Int16,  ClientEventTime DateTime,  SilverlightVersion1 UInt8,  SilverlightVersion2 UInt8,  SilverlightVersion3 UInt32,  SilverlightVersion4 UInt16,  PageCharset String,  CodeVersion UInt32,  IsLink UInt8,  IsDownload UInt8,  IsNotBounce UInt8,  FUniqID UInt64,  HID UInt32,  IsOldCounter UInt8,  IsEvent UInt8,  IsParameter UInt8,  DontCountHits UInt8,  WithHash UInt8,  HitColor FixedString(1),  UTCEventTime DateTime,  Age UInt8,  Sex UInt8,  Income UInt8,  Interests UInt16,  Robotness UInt8,  GeneralInterests Array(UInt16),  RemoteIP UInt32,  RemoteIP6 FixedString(16),  WindowName Int32,  OpenerName Int32,  HistoryLength Int16,  BrowserLanguage FixedString(2),  BrowserCountry FixedString(2),  SocialNetwork String,  SocialAction String,  HTTPError UInt16,  SendTiming Int32,  DNSTiming Int32,  ConnectTiming Int32,  ResponseStartTiming Int32,  ResponseEndTiming Int32,  FetchTiming Int32,  RedirectTiming Int32,  DOMInteractiveTiming Int32,  DOMContentLoadedTiming Int32,  DOMCompleteTiming Int32,  LoadEventStartTiming Int32,  LoadEventEndTiming Int32,  NSToDOMContentLoadedTiming Int32,  FirstPaintTiming Int32,  RedirectCount Int8,  SocialSourceNetworkID UInt8,  SocialSourcePage String,  ParamPrice Int64,  ParamOrderID String,  ParamCurrency FixedString(3),  ParamCurrencyID UInt16,  GoalsReached Array(UInt32),  OpenstatServiceName String,  OpenstatCampaignID String,  OpenstatAdID String,  OpenstatSourceID String,  UTMSource String,  UTMMedium String,  UTMCampaign String,  UTMContent String,  UTMTerm String,  FromTag String,  HasGCLID UInt8,  RefererHash UInt64,  URLHash UInt64,  CLID UInt32,  YCLID UInt64,  ShareService String,  ShareURL String,  ShareTitle String,  `ParsedParams.Key1` Array(String),  `ParsedParams.Key2` Array(String),  `ParsedParams.Key3` Array(String),  `ParsedParams.Key4` Array(String),  `ParsedParams.Key5` Array(String),  `ParsedParams.ValueDouble` Array(Float64),  IslandID FixedString(16),  RequestNum UInt32,  RequestTry UInt8) ENGINE = MergeTree PARTITION BY toYYYYMM(EventDate) SAMPLE BY intHash32(UserID) ORDER BY (CounterID, EventDate, intHash32(UserID), EventTime);

CREATE TABLE test.visits ( CounterID UInt32,  StartDate Date,  Sign Int8,  IsNew UInt8,  VisitID UInt64,  UserID UInt64,  StartTime DateTime,  Duration UInt32,  UTCStartTime DateTime,  PageViews Int32,  Hits Int32,  IsBounce UInt8,  Referer String,  StartURL String,  RefererDomain String,  StartURLDomain String,  EndURL String,  LinkURL String,  IsDownload UInt8,  TraficSourceID Int8,  SearchEngineID UInt16,  SearchPhrase String,  AdvEngineID UInt8,  PlaceID Int32,  RefererCategories Array(UInt16),  URLCategories Array(UInt16),  URLRegions Array(UInt32),  RefererRegions Array(UInt32),  IsYandex UInt8,  GoalReachesDepth Int32,  GoalReachesURL Int32,  GoalReachesAny Int32,  SocialSourceNetworkID UInt8,  SocialSourcePage String,  MobilePhoneModel String,  ClientEventTime DateTime,  RegionID UInt32,  ClientIP UInt32,  ClientIP6 FixedString(16),  RemoteIP UInt32,  RemoteIP6 FixedString(16),  IPNetworkID UInt32,  SilverlightVersion3 UInt32,  CodeVersion UInt32,  ResolutionWidth UInt16,  ResolutionHeight UInt16,  UserAgentMajor UInt16,  UserAgentMinor UInt16,  WindowClientWidth UInt16,  WindowClientHeight UInt16,  SilverlightVersion2 UInt8,  SilverlightVersion4 UInt16,  FlashVersion3 UInt16,  FlashVersion4 UInt16,  ClientTimeZone Int16,  OS UInt8,  UserAgent UInt8,  ResolutionDepth UInt8,  FlashMajor UInt8,  FlashMinor UInt8,  NetMajor UInt8,  NetMinor UInt8,  MobilePhone UInt8,  SilverlightVersion1 UInt8,  Age UInt8,  Sex UInt8,  Income UInt8,  JavaEnable UInt8,  CookieEnable UInt8,  JavascriptEnable UInt8,  IsMobile UInt8,  BrowserLanguage UInt16,  BrowserCountry UInt16,  Interests UInt16,  Robotness UInt8,  GeneralInterests Array(UInt16),  Params Array(String),  `Goals.ID` Array(UInt32),  `Goals.Serial` Array(UInt32),  `Goals.EventTime` Array(DateTime),  `Goals.Price` Array(Int64),  `Goals.OrderID` Array(String),  `Goals.CurrencyID` Array(UInt32),  WatchIDs Array(UInt64),  ParamSumPrice Int64,  ParamCurrency FixedString(3),  ParamCurrencyID UInt16,  ClickLogID UInt64,  ClickEventID Int32,  ClickGoodEvent Int32,  ClickEventTime DateTime,  ClickPriorityID Int32,  ClickPhraseID Int32,  ClickPageID Int32,  ClickPlaceID Int32,  ClickTypeID Int32,  ClickResourceID Int32,  ClickCost UInt32,  ClickClientIP UInt32,  ClickDomainID UInt32,  ClickURL String,  ClickAttempt UInt8,  ClickOrderID UInt32,  ClickBannerID UInt32,  ClickMarketCategoryID UInt32,  ClickMarketPP UInt32,  ClickMarketCategoryName String,  ClickMarketPPName String,  ClickAWAPSCampaignName String,  ClickPageName String,  ClickTargetType UInt16,  ClickTargetPhraseID UInt64,  ClickContextType UInt8,  ClickSelectType Int8,  ClickOptions String,  ClickGroupBannerID Int32,  OpenstatServiceName String,  OpenstatCampaignID String,  OpenstatAdID String,  OpenstatSourceID String,  UTMSource String,  UTMMedium String,  UTMCampaign String,  UTMContent String,  UTMTerm String,  FromTag String,  HasGCLID UInt8,  FirstVisit DateTime,  PredLastVisit Date,  LastVisit Date,  TotalVisits UInt32,  `TraficSource.ID` Array(Int8),  `TraficSource.SearchEngineID` Array(UInt16),  `TraficSource.AdvEngineID` Array(UInt8),  `TraficSource.PlaceID` Array(UInt16),  `TraficSource.SocialSourceNetworkID` Array(UInt8),  `TraficSource.Domain` Array(String),  `TraficSource.SearchPhrase` Array(String),  `TraficSource.SocialSourcePage` Array(String),  Attendance FixedString(16),  CLID UInt32,  YCLID UInt64,  NormalizedRefererHash UInt64,  SearchPhraseHash UInt64,  RefererDomainHash UInt64,  NormalizedStartURLHash UInt64,  StartURLDomainHash UInt64,  NormalizedEndURLHash UInt64,  TopLevelDomain UInt64,  URLScheme UInt64,  OpenstatServiceNameHash UInt64,  OpenstatCampaignIDHash UInt64,  OpenstatAdIDHash UInt64,  OpenstatSourceIDHash UInt64,  UTMSourceHash UInt64,  UTMMediumHash UInt64,  UTMCampaignHash UInt64,  UTMContentHash UInt64,  UTMTermHash UInt64,  FromHash UInt64,  WebVisorEnabled UInt8,  WebVisorActivity UInt32,  `ParsedParams.Key1` Array(String),  `ParsedParams.Key2` Array(String),  `ParsedParams.Key3` Array(String),  `ParsedParams.Key4` Array(String),  `ParsedParams.Key5` Array(String),  `ParsedParams.ValueDouble` Array(Float64),  `Market.Type` Array(UInt8),  `Market.GoalID` Array(UInt32),  `Market.OrderID` Array(String),  `Market.OrderPrice` Array(Int64),  `Market.PP` Array(UInt32),  `Market.DirectPlaceID` Array(UInt32),  `Market.DirectOrderID` Array(UInt32),  `Market.DirectBannerID` Array(UInt32),  `Market.GoodID` Array(String),  `Market.GoodName` Array(String),  `Market.GoodQuantity` Array(Int32),  `Market.GoodPrice` Array(Int64),  IslandID FixedString(16)) ENGINE = CollapsingMergeTree(Sign) PARTITION BY toYYYYMM(StartDate) SAMPLE BY intHash32(UserID) ORDER BY (CounterID, StartDate, intHash32(UserID), VisitID);

clickhouse-client --max_insert_block_size 100000 --query "INSERT INTO test.hits FORMAT TSV" < hits_v1.tsv
clickhouse-client --max_insert_block_size 100000 --query "INSERT INTO test.visits FORMAT TSV" < visits_v1.tsv
```

**Comment by alexey-milovidov** [comment]
Performance tests are Ok.

**Comment by alexey-milovidov** [comment]
> Stateful tests should be updated.

I'll do it.

**Comment by alexey-milovidov** [comment]
Now it gives somewhat wrong result for `firstSignificantSubdomain('magnet:yandex.abc')`.
It is considered to be Ok.

**Comment by alexey-milovidov** [comment]
It is definitely not the expected behaviour:

```
SELECT 
    domain(URL) AS domain, 
    URL
FROM test.hits 
WHERE (URL NOT LIKE '%://%') AND (domain != '')
LIMIT 10

┌─domain─┬─URL─────────┐
│ about  │ about:blank │
│ about  │ about:blank │
└────────┴─────────────┘
┌─domain─┬─URL───────────┐
│ tel    │ tel:904 53 00 │
└────────┴───────────────┘
┌─domain─────┬─URL─────────────────────────────────┐
│ mailto     │ mailto:kalinin.yapanorama@yandex.ru │
│ mailto     │ mailto:nikitina@reklama3.com        │
│ javascript │ javascript:this.name                │
│ javascript │ javascript:this.name                │
│ javascript │ javascript:this.name                │
│ javascript │ javascript:this.name                │
│ about      │ about:blank                         │
└────────────┴─────────────────────────────────────┘
```

**Comment by alexey-milovidov** [comment]
It should give correct results for both:

```
https://yandex.ru/ -> yandex.ru
yandex.ru:80 -> yandex.ru
mailto:hello@world -> 
magnet:test.test ->
```

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
.

**Comment by YiuRULE** [comment]
Ok thanks, I applied a patch for the required fix and up to date the tests.

I just have few updates, 5 tests with the `SELECT ... test.hits` have a match.

```https_edit.do?res=5&uinfo=ww-980-wh-660-fps-v-krasnoyaroslavy-ispytyvat-i-table_key=5&has_photelnoe => https_edit.do
https_edit.do?res=5&uinfo=ww-980-wh-660-fps-v-krasnoyaroslavy-ispytyvat-i-table_key=5&has_photelnoe => https_edit.do
http-www-1349-wh-957-fw-1038-03-28.0395806813113 => http-www-1349-wh-957-fw-1038-03-28.0395806813113
http-www-1349-wh-957-fw-1038-03-28.0395806813113 => http-www-1349-wh-957-fw-1038-03-28.0395806813113```

For this 4 tests, I cannot do that much things without doing a TLD lookup, should we also check the TLD here too ?

Also have an another case

```https_edit.do?id=1882117.jpg","can  => https_edit.do```

Should we also refuse this one due to some invalid characters ?

Otherwise the test case without an url seem to work now

**Comment by alexey-milovidov** [comment]
Now the tests queries became 30% slower:
https://clickhouse-test-reports.s3.yandex.net/5045/8b0dda39e4e268af282e77c2d24f89f04b3f64c6/performance_test/comparison_min_time_gcc_8.html

I have to either do some magic or just feel fine about it.

**Comment by alexey-milovidov** [comment]
@YiuRULE 
```
https_edit.do?res=5&uinfo=ww-980-wh-660-fps-v-krasnoyaroslavy-ispytyvat-i-table_key=5&has_photelnoe => https_edit.do
http-www-1349-wh-957-fw-1038-03-28.0395806813113 => http-www-1349-wh-957-fw-1038-03-28.0395806813113
http-www-1349-wh-957-fw-1038-03-28.0395806813113 => http-www-1349-wh-957-fw-1038-03-28.0395806813113
```
These results are Ok. We should not check the TLD.
These are fabricated URLs and current behaviour is totally fine (garbage in -> garbage out).

File: dbms/src/Functions/tl.sql [hunk 3]
```diff
@@ -0,0 +1,26 @@
+CREATE TABLE ip(
```
Comment (alexey-milovidov) [inline dbms/src/Functions/tl.sql]: Unrelated file?

**Comment by geldot** [comment]
Does this fix cutToFirstSignificantSubdomain() etc? Assuming the other functions call this one, having some tests for those too would be great (and if not we should fix those too maybe as a separate issue)

We have `cutToFirstSignificantSubdomain(concatAssumeInjective('https://',domain))` in our queries right now, looking forward to removing.

File: dbms/src/Functions/tl.sql (hunk 3 above)
Comment (YiuRULE) [reply dbms/src/Functions/tl.sql]: Indeed, thanks, I removed it

**Comment by YiuRULE** [comment]
@geldot Yep, the patch fix also your case, I forget to add test and just added some test for the function `cutToFirstSignificantSubdomain`, thanks !

**Comment by YiuRULE** [comment]
Does we need to make other changes ? would be glad to assist if needed :)

**Comment by YiuRULE** [comment]
Any news concerning this PR ?

**Comment by alexey-milovidov** [comment]
> Any news concerning this PR ?

We are currently working on this PR, trying to squeeze missing bits of performance. Results are optimistic.

**Pull request merged**